libzfs/src/zfs.rs

```diff
@@ -0,0 +1,430 @@
+// Copyright (c) 2018 Intel Corporation. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+extern crate libzfs_sys as sys;
+
+use std::io::Error;
+use std::ptr;
+use std::ffi::{CStr, CString};
+use nvpair;
+use libzfs_error::{LibZfsError, Result};
+use zprop_list::{ZProp, ZpropItem, ZpropList};
+
+#[derive(Debug, PartialEq)]
+pub struct Zfs {
+    raw: *mut sys::zfs_handle_t,
+}
+
+impl Zfs {
+    pub fn new(raw: *mut sys::zfs_handle_t) -> Zfs {
+        Zfs { raw }
+    }
+    pub fn name(&self) -> CString {
+        let s = unsafe { CStr::from_ptr(sys::zfs_get_name(self.raw)) };
+        s.to_owned()
+    }
+    pub fn user_props(&self) -> &mut nvpair::NvListRef {
+        unsafe {
+            let x = sys::zfs_get_user_props(self.raw);
+            nvpair::NvListRef::from_mut_ptr(x)
+        }
+    }
+    pub fn zfs_type(&self) -> sys::zfs_type_t {
+        unsafe { sys::zfs_get_type(self.raw) }
+    }
+    pub fn zfs_type_name(&self) -> CString {
+        let x = self.zfs_type();
+
+        let s = unsafe { CStr::from_ptr(sys::zfs_type_to_name(x)) };
+
+        s.to_owned()
+    }
+    pub fn prop_list(&self) -> Result<ZpropList> {
+        let mut prop_list_ptr: *mut sys::zprop_list_t = ptr::null_mut();
+
+        let code = unsafe {
+            sys::zfs_expand_proplist(
+                self.raw,
+                &mut prop_list_ptr,
+                sys::boolean::B_TRUE,
+                sys::boolean::B_TRUE,
+            )
+        };
+
+        match code {
+            0 => Ok(ZpropList::new(prop_list_ptr)),
+            x => Err(LibZfsError::Io(Error::from_raw_os_error(x))),
+        }
+    }
+    pub fn props(&self) -> Result<(Vec<ZProp>)> {
+        let buff_size = 319;
```

Comment:
What's the significance of 319?

Reply:
320 is the maximum size I found usable for the pre-allocated buffer that gets the property.

319 + null byte = 320.